### ISSUE TITLE
Improve performance of `object.mutate` by batching update handling

### ISSUE THREAD

**Issue by trusktr** [nasa/openmct#4648]
**Is your feature request related to a problem? Please describe.**

In APRES, it currently takes about 2 ms calling `objects.mutate` on a domain object. Perhaps we can reduce this overhead by deferring the mutation handling to a microtask (so it is still synchronous, but at the end of the current macrotask). This may be able to reduce the overhead from O(n) to O(1); f.e. 2ms total instead of 2ms per `mutate` call.

**Describe the solution you'd like**

I'm not sure exactly what it entails, but the idea is to batch mutation handlers, and finally trigger observation handlers at the end (I'm not sure if that's already the case with observation handlers, but if not, then this could be a breaking change requiring a major version bump, although it wouldn't be as breaking as the `Object.seal` idea in https://github.com/nasa/openmct/issues/4647).

**Describe alternatives you've considered**
N/A

**Additional context**

In APRES, property sets on domain objects are taking 4 ms, approximately half of which is from `objects.mutate` calls. The other half is from Vue 2 reactive setters and the solution for that is in https://github.com/nasa/openmct/issues/4647.

**Comment by akhenry** [comment]
This is an interesting idea, thanks. I suspect that there are some cases in our codebase that rely upon mutation handlers running _immediately_ upon mutation, but there probably aren't many cases like that. 

Mutation hasn't been a performance bottleneck for us, because it doesn't happen that often in our use cases, so we haven't focused on it like we have with telemetry processing. Is this causing performance issues that are noticeable to the user in APRES? If so, can you describe the use case a little?

**Comment by trusktr** [comment]
> Is this causing performance issues that are noticeable to the user in APRES? If so, can you describe the use case a little?

Yes indeed: APRES is relying on two things: mutating domain objects, and then relying on Vue reactivity to detect changes of those objects, in order to animate UI on mouse move for example.

For each mouse move (f.e. dragging an action in timeline) a mutation takes half the time. The other half is Vue 2 signaling reactivity to the Vue component to make it update. Total is about 4ms in our case, on my machine: 2ms mutating, 2ms triggering reactivity, per mouse move.

You're right that 2ms can be negligible for one-off use cases, but when it comes to animation (f.e. by animating numbers in domainObjects, or animating numbers that Vue 2 reactivity is triggered by such as those in our domainObjects) this is a really high number (where we have 16ms per frame for 60fps, and where part of that is the browser itself so the JavaScript has less than 16ms).

**Comment by akhenry** [comment]
Is mutation occurring on `mousemove`, or on `mouseup`? If it's `mousemove`, that's just way too many mutations, it's not intended to be performant at that rate of mutation. 

At the _very least_ any mouse move handlers should be throttled by `requestAnimationFrame`, but even that is going to result in a mutation frequency that is just just way too high if you're mutating on `mousemove`. 

I agree that mutation batching is probably a good approach, but I think that it should happen in the component, and not in the API. Doing it in the API involves fundamentally changing mutation from a synchronous event to an asynchronous event, which could potentially have a lot of knock on effects.

**Comment by trusktr** [comment]
> I think that it should happen in the component, and not in the API

Can you give an example of this? How do we set multiple properties of a domain object at once outside of the API, and then trigger a single reactive event?

> Doing it in the API involves fundamentally changing mutation from a synchronous

What if it is an option, disabled by default? F.e.

```js
objects.mutate(object, path, value, {batch: true})
```

or similar.

The downside of it not being in the API is: 

- how should the user do it?
- whatever the solution, each separate consumer has to implement the additional complexity

**Comment by trusktr** [comment]
Another idea is we can add a new `Proxy`-based API, so that we can just do this: `someDomainObjectProxyVersion.subObject.someProp = whatever` and it would default to batched reactivity.

**Comment by trusktr** [comment]
If we ever do add a proxy version, we can take advantage of existing libs for this.

For example, here's how to do it using [Solid.js](https://www.solidjs.com/), while keeping the event API:

```js
import {createEffect} from 'solid-js'
import {createMutable} from 'solid-js/store'

export function createDomainObject(obj) {
  const mutable = createMutable(obj) // wraps obj in a Proxy
  // ... any other tracking of created domain objects ...
  return mutable
}

// similar to objects.observe API
export function observe(obj, path, cb) {
  const pathParts = // ... parse path into array of path parts ... 

  let dispose
  createRoot(stop => {
    dispose = stop

    createEffect(() => {
      // read the path. This tracks dependencies for the effect, and the effect will re-run any time the path values change.
      let o = obj
      for (const part of pathParts) o = o[part]

      cb(o)
    })
  })
}
```

the user can use it like this:

```js
import {createDomainObject} from 'openmct'

const domainObject = createDomainObject({
  foo: {
    bar: 123
  }
})

observe(domainObject, 'foo.bar', (value) => {
  console.log('domainObject.foo.bar is ', value)
})
```

The idea of this `observe` function is that it would be similar to the existing `objsects.observe` API, but it is unnecessary if the user just uses Solid APIs directly. Here's the same user code, but simpler:

```js
import {createDomainObject} from 'openmct'
import {createEffect} from 'solid-js'

const domainObject = createDomainObject({
  foo: {
    bar: 123
  }
})

// The effect automatically re-runs when that path changes anywhere along it.
createEffect(() => {
  console.log('domainObject.foo.bar is ', domainObject.foo.bar)
})
```

Solid.js has a `batch` API built in. Suppose the user is observing a bunch of properties on a domain object:

```js
createEffect(() => {
  console.log('All the values are: ', obj.foo, obj.bar, obj.baz, obj.lorem, obj.ipsum)
})
```

then this code would normally cause the effect to run five times, once time per property set (synchronously):

```js
obj.foo = 1
obj.bar = 1
obj.baz = 1
obj.lorem = 1
obj.ipsum = 1
```

but the user can batch them like the following to cause the effect to run only once after updating all five properties:

```js
import {batch} from 'solid-js'

batch(() => {
  obj.foo = 1
  obj.bar = 1
  obj.baz = 1
  obj.lorem = 1
  obj.ipsum = 1
})
```

It is still synchronous, just that now it happens after the function passed to `batch` is done running synchronously. `batch` runs the function immediately. So, writing it like the following would cause the effect to run twice:

```js
batch(() => {
  obj.foo = 1
  obj.bar = 1
})

// the effect runs now

batch(() => {
  obj.baz = 1
  obj.lorem = 1
  obj.ipsum = 1
})

// the effect runs now
```

One of the reasons I love Solid.js is because it's state system is entirely decoupled from its UI components: in the above examples, we did not write a single UI component, but we were able to use Solid primitives to create reactive state and react to state changes. In contrast, for example, we can't do this with React: in React, reactive state can only be created _inside_ of a component.

If we want to take a domainObject and show a value within some UI, easy:

```js
export const obj = createDomainObject({foo: 123, ...})
```
```js
import {obj} from './obj'

export function SomeUI() {
  return <h1>obj.foo is: {obj.foo}</h1>
}
```

If we adopt the Proxy trick, I'd use Solid rather than make something from scratch. They have a passionate team specializing in this sort of thing, advancing this area, winning at benchmarks ([compare it](https://krausest.github.io/js-framework-benchmark/current.html) to React, Vue, Angular, Svelte, etc).

**Comment by akhenry** [comment]
> Can you give an example of this? How do we set multiple properties of a domain object at once outside of the API, and then trigger a single reactive event?

Maybe we're talking about different problems. There's no explicit support for batching of the type you've described. You _could_ do something like:
```
openmct.objects.mutate(domainObject, "dimensions", {width: 10, height:20});
```
Instead of 
```
openmct.objects.mutate(domainObject, "dimensions.width", 10);
openmct.objects.mutate(domainObject, "dimensions.height", 20);
```
but that only solves for batching of properties on the same path.

What I was suggesting is that you can update the _view model_ at whatever rate you like to give the user real-time feedback, and only periodically update the persistence model (ie. the domainObject).

An analogy I can think of is in our tables. A user can drag a column around to resize or reorder it, and that is reflected in real-time because properties on the Vue model are changing in real-time, but the table domain object itself is only updated when the user releases the mouse button.

**Comment by trusktr** [comment]
> but that only solves for batching of properties on the same path.

Good to know.

What I would do is, rather than make the user think about what to batch, I would make the observation mechanism always batched by default. What would happen is this:

- the user sets properties (might be on different paths, doesn't matter)
- during this process, when any property any level deep is set, an observation is queued for later in a microtask
- after the current synchronous turn of the user's current task, the microtask will run the observation only _once_.

The behavior would be like this. This is the current:


```js
openmct.objects.mutate(domainObject, "dimensions.width", 10); // objects.observe callback runs before this line finishes
openmct.objects.mutate(domainObject, "dimensions.height", 20); // objects.observe callback runs again before this line finishes
```

New behavior would be:

```js
openmct.objects.mutate(domainObject, "dimensions.width", 10); // no observe callback happens yet
openmct.objects.mutate(domainObject, "dimensions.height", 20); // no observe callback happens yet

// ...later, in the next microtask, the objects.observe callback is called only once...
```

In other projects of mine, this sort of pattern fixes bugs and improves performance:

- it prevents intermediary states where between multiple property sets the final state is not complete
- reduces the number of reactions happening unnecessarily

Here are live examples of the concept (using Solid APIs):

This first example has an issue due to an effect running in between writes to reactive variables, and inconsistent state causes the visual scene to be removed and then re-created. The problem happens to be very obvious because I chose to render something that takes a while to load (due to it being WebGL, having to re-upload to GPU memory):

https://playground.solidjs.com/?hash=1025549064&version=1.3.13

This next example shows how a user can solve it by using additional cognitive overhead from having to think about using the `batch` API. Notice that the scene no longer unloads and reloads every second:

https://playground.solidjs.com/?hash=-394158961&version=1.3.13

Finally, this next example shows the concept of a deferred-batched reactive effect just like I described in the bullet points above. Note the user did not have to worry about how to batch things, they just write code like the first example, and it simply works:

https://playground.solidjs.com/?hash=-54762955&version=1.3.13

It'll be nice having this for `objects.observe`.

Side note, the recent release of `React` 18 fully migrated to this approach, finally.

The [Qt](http://qt.io/) framework upgraded to have similar deferred-batched options in v5 and 6 a few years back.

**Comment by trusktr** [comment]
The main idea is that, if you depend on something, it doesn't matter _when_ the reaction happens. What matter is that it will happen, and at that point, you react to it. Taking advantage of the it-doesnt-matter-when concept, this makes it ok to batch-defer, because as a consumer we really don't know when our reaction callback will run, and we shouldn't depend on that because it'll make our code brittle; and if we do need to depend on it, we should do so by depending on further reactive variables such that those further reactions run whenever they do without us caring, rather than us trying to guess.

**Comment by trusktr** [comment]
> What I was suggesting is that you can update the view model at whatever rate you like to give the user real-time feedback, and only periodically update the persistence model (ie. the domainObject).

For the persistence layer, yeah throttling can make great sense. I had to do that with that `first-person-shooter` concept to prevent it from DoS'ing each player's browser tab. :D

I believe that deferred-batched should be the default for data in any system, along with lazy-getters for all derived variables in that a would-be-deferred-batched value is needed before it would otherwise be updated (the early access erases the future update from needing to happen). 

This is all noting that deferred-batched (or I should say, microtask-deferred-batched) is still synchronous (microtasks are synchronously executed in the macrotask they originate from, it is only code re-ordering). This doesn't prevent someone from making a synchronous loop while continually updating a variable every time it changes, and it would block the CPU.

This microtask-deferred-batching doesn't prevent the app from having an immediate state, just that code re-ordering prevent unnecessary amounts of code execution in the same overall tick (one tick having a single macrotask and any number of microtasks).

In contrast, an async throttle for persistence would be slower, with actual timing (like setTimeout) in between event loop macrotasks, and would not be necessarily be in sync with the immediate state of the app at any point in time.

Note also that the browser paint can only happen between one macrotask, and the next: so if we microtask-defer-batch reactivity reactions, that does not mean that the reactions will be interleaved with paint cycles and cause unexpected delayed renderings on screen; microtask-deferred-batched reactions will settle synchronously, then paint will happen.

I believe this will lead to the most performance, while avoiding inconsistent state.
